Gutenboarding: Display FSE design using seedlet-blocks

noahtallen opened the pull request:
This resolves #43288, since you will be able to activate seedlet-blocks internally via Gutenboarding in staging environments. This also resolves #43285, which adds a design which when chosen, will activate FSE on the new site.

This diff is the companion on wpcom and has already been deployed: D45075-code.

### Changes proposed in this Pull Request
* Add `is_fse` flag and `seedlet-blocks` theme to a design
* Fix issue where selecting a premium plan did not redirect you to the site editor
* Added selector for FSE site now that there are multiple ways to enter the flow
* Instead of signup flow option, use new `enabled_fse` option to create the site. This way, the backend still considers the ultimate flow to be gutenboarding. (see code-D45075 as well.)
* When the `gutenboarding/site-editor` flag is set, convert every design to use `is_fse` and `seedlet-blocks` as the theme.

~This is currently blocked on #43415, which needs deployed so that the seedlet demo site begins working again. Once that is working, we can correctly generate previews. Currently previews do not work because the demo site does not render a post content block:~ (Previews are now working.)

~Second, we need the `is_edge` config option in #42354 to restrict the design to certain environents. Otherwise this will start working in prod.~ (This has been completed.)

### Known issues:
* Font options do not work with the `seedlet-blocks` theme.
* When loading the site editor for the first time, the incorrect template is used for the homepage, meaning that you do not see the post content of your selected design. This is fixed by https://github.com/WordPress/gutenberg/pull/23186, which will arrive on dotcom in the coming weeks.

### Testing instructions
* (if you are sandboxed, make sure it is up to date. being sandboxed is no longer a requirement)
* Go through `http://calypso.localhost:3000/new`
* Select the first design option (named "FSE Test Design")
* Select the premium plan
* Verify that you are redirected to the site editor after getting the premium plan
* Verify that the seedlet theme is active on your site.
* Verify that the `gutenberg-edge` sticker is applied to your site.
* Verify that the homepage post_content matches the Cassel page pattern (it should _not_ match the seedlet-blocks default homepage content.)

Ålso test:
* Use http://calypso.localhost:3000/new?flags=gutenboarding/site-editor
* Every design option should have the seedlet-blocks theme set and the previews should look correct.
* Finishing the flow should land in the site editor, with the selected design's page pattern showing on the front end.

noahtallen committed:
Use seedlet-blocks all the time if FSE is enabled

simison commented:
What do you think about overriding this at here:

https://github.com/Automattic/wp-calypso/blob/c116cb0a2c114a39a3670f0df6d49c6653bced17/client/landing/gutenboarding/available-designs.ts#L17

And using script to generate new thumbnails with Seedlet theme: 

https://github.com/Automattic/wp-calypso/blob/master/bin/generate-gutenboarding-design-thumbnails.js

That way everything should fall into place nicely, including site creation with correct theme.

You'll need theme demo site for the preview mechanism to work tho!

noahtallen commented:
I definitely like that idea! I think theme demo site is the missing part. if that is available, will it work for template API and for mshots?

simison commented:
Yep!

noahtallen committed:
Use separate FSE design config file

ockham commented:
> I definitely like that idea! I think theme demo site is the missing part. if that is available, will it work for template API and for mshots?

@ianstewart Can you help us set up a demo site for seedlets-blocks?

noahtallen commented:
> What do you think about overriding this at here:

done. I copied the existing design file and changed the theme property to `seedlet-blocks` and also changed the `src` to reference seedlet instead of the theme it had set.

ianstewart commented:
> Can you help us set up a demo site for seedlets-blocks?

Does it need to 1) involve the showcase in any way? Or 2) does it just need to be _any_ site on WP.com with any theme that has Headstart annotations. 

If it's number one … we might need a way to hide the theme from non-automatticians since it's not ready for launch. If it's number two … for sure.

noahtallen commented:
@ianstewart I don't know... it needs to work for Gutenboarding. So with mshots and the template API.

The theme activates correctly via the API already :) The only time the showcase appears to be broken right now is if the theme is already active on the site (because its preview image doesn't show up)

> does it just need to be any site on WP.com with any theme that has Headstart annotations.

This sounds true? I'm not sure what this distinction is.

ianstewart commented:
I set up a temporary demo site and generated temporary Headstart annotations as well.

https://seedletblocksdemo.wordpress.com/

The front-end doesn't look as expected currently but I believe that's just a known issue with our demo templates in the site editor on WP.com.

noahtallen committed:
Include fse design in thumbnail generation

add seedlet-theme to json

noahtallen commented:
TODO: integrate with changes in #42354

ramonjd commented:
> TODO: integrate with changes in #42354

You might get there first :P

noahtallen commented:
The problem with previews in this PR is related to two issues:

The seedlet demo site has some kind of template issue and doesn't render a static homepage. To fix:

    a) deploy #43334
    b) possibly wait for new Gutenberg plugin version to match latest bug fixes from yesterday
    c) update the demo site via site editor to fix the template issue whatever it is
    d) run preview update script again here
    e) cross fingers and hope it all works

Additionally, it seems that the seedlet blocks theme is not using gutenboarding headstart content or whatever shenanigans occur there, so all the content is started in _classic editor content_ mode 🕺

noahtallen commented:
> You might get there first 

it looks so close! :D

ramonjd commented:
> seedlet blocks theme is not using gutenboarding headstart content or whatever shenanigans occur there, so all the content is started in classic editor content mode

Looking at the seedletblocksdemo site, it appears seedlet blocks is its own theme. I suspect that there are no headstart annotation for headstart to use so it's using some old fallback. (?)

You could try tricking headstart into using a different annotation by passing it a vertical slug. I haven't tested it, but I'm talking about making it use `m1.json` etc that you'll find under `nux-helpers`.

simison commented on client/landing/gutenboarding/available-designs.ts:
```diff
@@ -9,12 +9,15 @@ import { addQueryArgs } from '@wordpress/url';
 }
 
-const availableDesigns: Readonly< AvailableDesigns > = availableDesignsConfig;
+const availableDesigns: Readonly< AvailableDesigns > = isEnabled( 'gutenboarding/site-editor' )
```
I agree this is probably best direction.

In intermediate, to me it seems totally fine to also just override `theme` value in _all_ designs when reading the config. That would be very little code changes until we know more, and until we have this coming from an API anyway.

You could switch Gutenboarding to FSE mode with a URL like `wpcalypso.wordpress.com/new?flags=gutenboarding/site-editor` and everything would look as before, but theme would end up being different and final redirection URL would be different.

No strong opinion on this tho and `is_fse` seems fine too  (or opposite with `is_legacy`? Everything will be FSE eventually.).

To me separate config file is least preferable at this point, as that seems like a maintenance trouble.

noahtallen replied:
> it seems totally fine to also just override theme value in all designs when reading the config

I think this will cause issues with design preview if I'm not mistaken. (Since the screenshots would not be generated via that script)

noahtallen replied:
these changes are done now, other than integrating with `is_edge`

simison replied:
> I think this will cause issues with design preview if I'm not mistaken. (Since the screenshots would not be generated via that script)

You'd generate another set of preview thumbnails with `gutenboarding/site-editor` flag active locally; you'd end up with different set of files where unique URL has the new theme slug and they'd work just fine alongside other thumbnail files.

noahtallen replied:
Ah good idea. I'm not sure if we should go this extra step... I wonder if it would be confusing to switch out all of the themes if the flag is active?

Either way, it could be easy to add.

Currently, with the fag active:
- site editor is enabled
- gutenberg edge is enabled
- redirect to site editor
- theme is matched to whatever design you chose

Currently, without the flag, you can choose an FSE design and all of the above happens, plus the theme is set to seedlet-blocks.

I could see why overriding all designs with the flag might be helpful for now, I'll look into that

noahtallen replied:
@simison This is done now. I refactored `getAvailableDesigns` so that we could reuse it in the thumbnail script, and I also generated thumbnails for every design with seedlet-theme set.

noahtallen commented:
> I suspect that there are no headstart annotation for headstart to use

I think @ianstewart was able to fix this! 💙

ianstewart commented:
> I think @ianstewart was able to fix this! 💙

Yep. I made and deployed one. If anyone runs into an issue just ping me in Slack. I'm familiar enough with Headstart commands to make updates quickly.

noahtallen commented:
btw, generating working FSE designs is blocked on #43415

noahtallen committed:
Use fse-specific design instead of fse config

noahtallen committed:
Add enable_fse param to create site

noahtallen committed:
Remove changes to generate design thumbnails

noahtallen committed:
Add is_alpha property to FSE design

simison commented on client/landing/gutenboarding/stores/onboard/actions.ts:
```diff
@@ -122,10 +125,9 @@ export function* createSite(
-			theme: `pub/${ selectedDesign?.theme || 'twentytwenty' }`,
+			site_creation_flow: 'gutenboarding',
+			enable_fse: shouldEnableFse,
+			theme: `pub/${ selectedDesign?.theme || defaultTheme }`,
```
Odd that `defaultTheme` didn't need `pub/`? 🤔

noahtallen replied:
This should prefix both selected design and defaultTheme with `pub/`, right?

vindl replied:
Right.

simison commented on static/page-templates/page-templates.json:
```diff
@@ -163,6 +163,10 @@
+		},
+		{
+			"title": "seedlet-blocks",
+			"demo": "https://seedletblocksdemo.wordpress.com/"
```
OK to add, but apparently it isn't used much anywhere: https://github.com/Automattic/wp-calypso/pull/42354#discussion_r426955475

simison replied:
Yasss, of course. My bad :-)

noahtallen committed:
Add more fse designs

noahtallen commented:
Note: I had to `chmod +x` the thumbnail scripts

noahtallen commented:
Previews work now:
<img width="2033" alt="Screen Shot 2020-06-18 at 2 25 41 PM" src="https://user-images.githubusercontent.com/6265975/85073473-b1c22b80-b16f-11ea-8421-c1c801bbe7a2.png">
<img width="1511" alt="Screen Shot 2020-06-18 at 2 26 15 PM" src="https://user-images.githubusercontent.com/6265975/85073475-b2f35880-b16f-11ea-9828-0bec9c2edbeb.png">

Note: font selections _do not work_ with seedlet-blocks theme.

vindl reviewed (approved):
Worked great for me! 🚀

noahtallen committed:
Remove seedlet blocks from page templates

ockham commented on client/landing/gutenboarding/get-available-designs.js:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * This function reads the available designs from the config. This is written
+ * in raw JS so that the node script for generating previews can use this to
```
```suggestion
 * in CommonJS so that the node script for generating previews can use this to
```

Nitpick, but I find it a bit clearer than 'raw JS' :grimacing:

noahtallen replied:
you are very right

ockham commented on client/landing/gutenboarding/hooks/use-on-site-creation.ts:
```diff
@@ -107,9 +107,13 @@ export default function useOnSiteCreation() {
 					const redirectionUrl = isEcommercePlan
 						? `/checkout/${ newSite.site_slug }?preLaunch=1&isGutenboardingCreate=1`
-						: `/checkout/${ newSite.site_slug }?preLaunch=1&isGutenboardingCreate=1&redirect_to=%2Fblock-editor%2Fpage%2F${ newSite.site_slug }%2Fhome`;
+						: `/checkout/${ newSite.site_slug }?preLaunch=1&isGutenboardingCreate=1&redirect_to=%2F${ editorUrl }`;
```
Nice! :+1:

ockham commented on client/landing/gutenboarding/stores/onboard/selectors.ts:
```diff
@@ -25,3 +26,7 @@ export const getSelectedDomain = ( state: State ) => state.domain;
 export const getSelectedSiteTitle = ( state: State ) => state.siteTitle;
 export const getDomainSearch = ( state: State ) =>
 	state.domainSearch || getSelectedSiteTitle( state );
+export const getIsFseSite = ( state: State ): boolean => {
```
For boolean selectors, I think we have kind of a convention in Calypso to drop  the `get` prefix, and just start with `is` -- basically like the question they ask:

```suggestion
export const isFseSite = ( state: State ): boolean => {
```

noahtallen replied:
cool! now that I think of it, we can probably remove this selector since we can always use the design as the source of truth

noahtallen committed:
Remove unneeded FSE selector

ockham commented:
Thanks! This is great!

> * Instead of signup flow option, use new `enabled_fse` option to create the site. This way, the backend still considers the ultimate flow to be gutenboarding. (see code-D45075 as well.)

I :heart: this in particular -- your choice to use site options is much better (as it preserves other Gutenboarding-specific customizations) than mine was.

> * When the `gutenboarding/site-editor` flag is set, convert every design to use `is_fse` and `seedlet-blocks` as the theme.

I'm a bit skeptical about this -- IIUC, it will also require re-running the thumbnails generation script? I think that's easy to miss -- I'd rather make the available designs JSON the complete source of truth; for a given available designs JSON (and no changes to the designs themselves), running the generator script should reproducibly yield the same result. If we want seedlet-based themes, we'll need to add them to the JSON -- I think that's a valid requirement.

ockham commented:
> I'm a bit skeptical about this -- IIUC, it will also require re-running the thumbnails generation script? I think that's easy to miss -- I'd rather make the available designs JSON the complete source of truth; for a given available designs JSON (and no changes to the designs themselves), running the generator script should reproducibly yield the same result. If we want seedlet-based themes, we'll need to add them to the JSON -- I think that's a valid requirement.

(The level of indirection and number of wrapped generator scripts is IMO a symptom of this.)

ockham commented on client/landing/gutenboarding/get-available-designs.js:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * This function reads the available designs from the config. This is written
+ * in CommonJS so that the node script for generating previews can use this to
```
The downside is that this is untyped, and currently producing a [TS error](https://app.circleci.com/pipelines/github/Automattic/wp-calypso/64315/workflows/c1d4e7fd-464a-4109-a9e1-61513621f9b0/jobs/780314/steps).

I'd opt for removing the `forceFSE` option ([see](https://github.com/Automattic/wp-calypso/pull/43349#issuecomment-646346637)), which will probably allow us to drop this script (and re-inline the remaining logic into where it was before) -- this should get us rid of the TS error.

noahtallen commented:
Though I do agree with you about the level of indirection, I think it practically works:

> it will also require re-running the thumbnails generation script

Ah I see what you mean. I fixed this issue in, c5f3f2d. So the generation script will work with all possible design options.

Anyways, the motivation for this is that if you have the flag set, you'll end up with a site editor, but no block based theme and no demo templates. So in that scenario, I think you'll have a broken site.

We could get rid of the flag? 😅

ockham commented:
> Though I do agree with you about the level of indirection, I think it practically works:
> 
> > it will also require re-running the thumbnails generation script
> 
> Ah I see what you mean. I fixed this issue in, [c5f3f2d](https://github.com/Automattic/wp-calypso/commit/c5f3f2db9fe5c5fd86246b48ffbc598280ebc8c5). So the generation script will work with all possible design options.
> 
> Anyways, the motivation for this is that if you have the flag set, you'll end up with a site editor, but no block based theme and no demo templates. So in that scenario, I think you'll have a broken site.
> 
> We could get rid of the flag?

Oh, okay, that rationale was lost on me. Well could we hide the non-block based themes if the flag is set?

noahtallen committed:
Filter FSE designs if flag is set

noahtallen commented:
> Well could we hide the non-block based themes if the flag is set?

Based on this, I've changed it so that enabling the flag filters the designs to be only the FSE ones. The FSE designs still show up normal horizon/wpcalypso, but the flag just removes the other ones. That way, any option the user chooses with the flag set will get them to an FSE site.

ockham commented on client/landing/gutenboarding/stores/onboard/selectors.ts:
```diff
@@ -2,6 +2,7 @@
  * Internal dependencies
  */
 import type { State } from './reducer';
+import { isEnabled } from '../../../../config';
```
Leftover?

```suggestion
```

noahtallen committed:
Remove stale import

ockham commented:
Thanks a lot for the latest round of changes that removed the `forceFSE` option, as discussed in Slack (p1592521647367900-slack-cylon)!

Code-wise, this is looking great now. One more question, can we remove some of the regenerated thumbnails that are part of this PR? I assume we only need the `reynolds-fse` and `vesta-fse` ones now?

noahtallen commented:
> can we remove some of the regenerated thumbnails that are part of this PR? I assume we only need the reynolds-fse and vesta-fse ones now

sure :)

noahtallen committed:
Add FSE design previews

noahtallen commented:
>  I assume we only need the reynolds-fse and vesta-fse ones now?

done.

ockham commented:
I just tested this locally _without_ any of the new feature flags, but was shown the new FSE designs in the design selector:

![image](https://user-images.githubusercontent.com/96308/85082972-a94e0d00-b1d0-11ea-9a21-d6028a0ee740.png)

noahtallen commented:
Yes, that's intentional :) We can change that behavior if we want to. Note that these designs are also "alpha" which means they won't show up on prod ;)

noahtallen committed:
Hide FSE designs normally

noahtallen commented:
> Yes, that's intentional

I'm walking back on this now. Since the site editor isn't totally usable right now due to a bug in Gutenberg (the wrong template is used for the homepage in the editor), I'm fine hiding them by default for the time being.

Addison-Stavlo reviewed (approved):
Tested this and it works well!  Awesome stuff!

Both creation flows worked as expected (with and without the flag).  And now that the flag is required the designs do not show up without it.

noahtallen commented:
Feel free to continue improving this flow if needed :) Props to the gutenboarding team for making it so easy to work with!
